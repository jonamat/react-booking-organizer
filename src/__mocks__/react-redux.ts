import { Customer, Employee, Reservation, Service } from '../types';

export const __mockCustomers: Array<Customer> = [
    {
        id: 'customer-FAKE-ID1',
        name: 'Customer name',
        surname: 'Surname',
    },
    {
        id: 'customer-FAKE-ID2',
        name: 'Other name',
        surname: 'Other surname',
    },
];

export const __mockEmployees: Array<Employee> = [
    {
        id: 'employee-FAKE-ID1',
        name: 'Employee name 1',
    },
    {
        id: 'employee-FAKE-ID2',
        name: 'Employee name 2',
    },
];

export const __mockServices: Array<Service> = [
    {
        id: 'service-FAKE-ID1',
        name: 'fake service 1',
        averageDuration: 30,
    },
    {
        id: 'service-FAKE-ID2',
        name: 'fake service 2',
        averageDuration: 45,
    },
];

export const __mockReservations: Array<Reservation> = [
    {
        id: 'reservation-FAKE-ID1',
        date: new Date('1 Jan 2000 10:00 GMT+0100'),
        serviceId: __mockServices[0].id,
        customerId: __mockCustomers[0].id,
        employeeId: __mockEmployees[0].id,
        notes: 'something',
    },
    {
        id: 'reservation-FAKE-ID2',
        date: new Date('1 Jan 2040 10:00 GMT+0100'),
        serviceId: __mockServices[1].id,
        customerId: __mockCustomers[0].id,
        employeeId: __mockEmployees[1].id,
        notes: 'something2',
    },
];

export let useStore = () => ({
    getState: () => ({
        database: {
            customers: __mockCustomers,
            employees: __mockEmployees,
            services: __mockServices,
            reservations: __mockReservations,
        },
    }),
});

export const __setUseStore = (value) => (useStore = value);
export const useDispatch = () => undefined;
export const connect = () => (component) => component;
