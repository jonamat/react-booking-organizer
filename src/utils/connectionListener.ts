import { isOffline } from '../redux';

/**
 * Trigger a state change on connection changes
 * @param dispatch Redux dispatch function
 */
const connectionListener = (dispatch) => {
    // First check
    dispatch(isOffline(!navigator.onLine));

    window.addEventListener('online', () => dispatch(isOffline(false)));
    window.addEventListener('offline', () => dispatch(isOffline(true)));
};

export default connectionListener;
