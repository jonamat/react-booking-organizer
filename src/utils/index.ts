// Utils
export { default as isEmployeeBusy } from './isEmployeeBusy';
export { default as isEmployeeOnDuty } from './isEmployeeOnDuty';
export { default as generateClosestValidDate } from './generateClosestValidDate';
export { default as getReservationsByDate } from './getReservationsByDate';
export { default as getReservationStatus } from './getReservationStatus';

export { default as formatMinutes } from './formatMinutes';
export { default as hourMinuteToReadableString } from './hourMinuteToReadableString';
export { default as translateWeekDays } from './translateWeekDays';

export { default as databaseSync } from './databaseSync';
export { default as connectionListener } from './connectionListener';

// Auth system
export { default as login } from './login';
export { default as logout } from './logout';

// Delete database elements
export { default as deleteCustomer } from './deleteCustomer';
export { default as deleteReservation } from './deleteReservation';
export { default as deleteEmployee } from './deleteEmployee';
export { default as deleteService } from './deleteService';

// Add database elements
export { default as addCustomer } from './addCustomer';
export { default as addEmployee } from './addEmployee';
export { default as addReservation } from './addReservation';
export { default as addService } from './addService';

// Edit database elements
export { default as editCustomer } from './editCustomer';
export { default as editEmployee } from './editEmployee';
export { default as editReservation } from './editReservation';
export { default as editService } from './editService';
