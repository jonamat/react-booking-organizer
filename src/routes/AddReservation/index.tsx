import React, { FormEvent, ReactNode, FC, useRef, useState, useEffect } from 'react';
import { useStore, useDispatch } from 'react-redux';
import { toast } from 'react-toastify';
import queryString from 'query-string';
import { KeyboardDateTimePicker } from '@material-ui/pickers';
import { MaterialUiPickersDate } from '@material-ui/pickers/typings/date';
import { RouteComponentProps } from 'react-router';
import {
    Typography,
    TextField,
    Button,
    Divider,
    FormControl,
    InputLabel,
    Select,
    MenuItem,
    CircularProgress,
} from '@material-ui/core';
import Autocomplete from '@material-ui/lab/Autocomplete';

import { Customer, Service, Employee, RootState } from '../../types';
import { closeDialog, openDialog } from '../../redux';
import { MIN_STEP, PROGRESSIVE_MINS_WITHIN_HOUR, TIME_DATE_FORMAT } from '../../config';
import { bindActionCreators } from 'redux';
import { generateClosestValidDate, addReservation } from '../../utils';
import Alert from '../../components/Alert';
import useStyles from './style';

interface ParsedInfo {
    customer?: Customer; // Autocomplete needs the whole object
    employeeId?: string;
    date?: Date;
}

const AddReservation: FC<RouteComponentProps> = ({ history, location }) => {
    const classes = useStyles();
    const inputForm = useRef<HTMLFormElement>(null);
    const {
        database: { customers, employees, services, reservations },
    } = useStore<RootState>().getState();
    const actions = bindActionCreators({ openDialog, closeDialog }, useDispatch());

    if (!customers || !employees || !services || !reservations)
        throw new Error('Internal error - Resources loading logic broken');

    const [loading, setLoading] = useState<boolean>(false);
    const [customer, setCustomer] = useState<Customer | null>(null);
    const [employeeId, setEmployeeId] = useState<string | null>(employees[0].id);
    const [selectedDate, setSelectedDate] = useState<MaterialUiPickersDate>(generateClosestValidDate());

    useEffect(() => {
        /* Check & validate search query if provided */
        const query = queryString.parse(location.search);
        const parsed: ParsedInfo = {};

        // Get customer
        if (query.customerId) {
            const parsedCustomer = customers.find((customer) => customer.id === query.customerId);
            if (!parsedCustomer) toast.error('Cliente non presente nel database');
            else parsed.customer = parsedCustomer;
        }

        // Get employee
        if (query.employeeId) {
            if (!employees.find((employee) => employee.id === query.employeeId))
                toast.error('Dipendente non presente nel database');
            else parsed.employeeId = query.employeeId as string;
        }

        // Get date
        if (query.hour && query.date) {
            /* Query format date=yyyy-mm-dd hour=HH:MM military time */
            const date = new Date(query.date + 'T' + query.hour);

            if (isNaN(date.getTime())) {
                toast.error('Formato data e ora non regolari');
            } else if (!PROGRESSIVE_MINS_WITHIN_HOUR.includes(date.getMinutes())) {
                toast.error(`L'orario indicano non rispetta lo scaglionamento orario dell'agenda.
                Utilizza intervalli di ${MIN_STEP} minuti.`);
            } else if (date.getTime() < Date.now()) {
                toast.warn(`Attenzione! Stai assegnando questo appuntamento nel passato`);
                parsed.date = date;
            } else parsed.date = date;
        }

        // Update local state to pre fill the fields
        if (parsed.customer) setCustomer(parsed.customer);
        if (parsed.date) setSelectedDate(parsed.date);
        if (parsed.employeeId) setEmployeeId(parsed.employeeId);
    }, []);

    const handleSubmit = (event: FormEvent<HTMLFormElement>, forceAssignment?: boolean) => {
        event.preventDefault();
        if (!inputForm.current || !selectedDate) return;

        // Disable user interactions
        setLoading(true);

        addReservation(
            new FormData(inputForm.current),
            customer,
            selectedDate,
            {
                employees,
                services,
                reservations,
            },
            forceAssignment,
        )
            .then((reservation) => {
                toast.success(
                    'Aggiunto nuovo appuntamento per ' + reservation.date.toLocaleDateString('it-IT', TIME_DATE_FORMAT),
                );
                history.push(`/reservations?date=${reservation.date.toISOString().slice(0, 10)}`);
            })
            .catch((error: Error) => {
                // If employee is not on duty that day, ask for confirmation and calls recursively the fn
                if (error.message === 'EMPLOYEE_NOT_ON_DUTY') {
                    actions.openDialog({
                        title: `Il dipendente selezionato non Ã¨ in servizio ${selectedDate.toLocaleDateString(
                            'it-IT',
                            TIME_DATE_FORMAT,
                        )}.`,
                        content: (
                            <Alert
                                cancelText="Annulla"
                                onCancel={actions.closeDialog}
                                confirmText="Conferma"
                                onConfirm={() => {
                                    actions.closeDialog();

                                    // Re call the fn forcing the assignment
                                    handleSubmit(event, true);
                                }}
                                contentText="Vuoi assegnare comunque l'appuntamento?"
                            />
                        ),
                    });
                } else {
                    toast.error(error.message);
                    setLoading(false);
                }
            });
    };

    // Controlled components handlers
    const handleChangeDate = (selectedDate: MaterialUiPickersDate) => setSelectedDate(selectedDate);
    const handleChangeCustomer = (_e: React.ChangeEvent<any>, customer: Customer | null) => setCustomer(customer);
    const handleChangeEmployeeId = (event) => setEmployeeId(event.target.value);

    // Case: user hasn't yet added services, customers or employees
    if (!services.length || !customers.length || !employees.length)
        return (
            <div className={classes.container} data-testid="add-reservation">
                <Typography variant="body1">
                    Devi aggiungere clienti, dipendenti e servizi prima di poter fissare appuntamenti
                </Typography>
            </div>
        );

    return (
        <div className={classes.container} data-testid="add-reservation">
            <Typography variant="h6">Aggiungi appuntamento</Typography>
            <Divider className={classes.divider} />
            <form className={classes.form} ref={inputForm} onSubmit={handleSubmit} data-testid="form">
                <Autocomplete
                    classes={{
                        root: classes.field,
                        popupIndicator: classes.autocompleteIndicators,
                        clearIndicator: classes.autocompleteIndicators,
                    }}
                    value={customer}
                    options={customers}
                    getOptionLabel={(customer: Customer): string =>
                        customer.name +
                        (customer.surname ? ' ' + customer.surname : '') +
                        (customer.additionalIdentifier ? ' ' + customer.additionalIdentifier : '')
                    }
                    autoHighlight
                    noOptionsText="Cliente non trovato"
                    renderInput={(params): ReactNode => (
                        <TextField
                            {...params}
                            variant="outlined"
                            InputProps={{
                                ...params.InputProps,
                            }}
                            inputProps={{
                                ...params.inputProps,
                                'data-testid': 'customer',
                            }}
                            placeholder="Cliente"
                        />
                    )}
                    onChange={handleChangeCustomer}
                />

                <KeyboardDateTimePicker
                    className={classes.field}
                    variant="dialog"
                    inputVariant="outlined"
                    format="dd/MM/yyyy HH:mm"
                    ampm={false}
                    label="Data *"
                    invalidDateMessage="Data non valida"
                    minutesStep={15}
                    value={selectedDate}
                    onChange={handleChangeDate}
                    inputProps={{
                        'data-testid': 'date',
                    }}
                />

                <FormControl variant="outlined" className={classes.field} required>
                    <InputLabel className={classes.label} id="servicesIdLabel">
                        Servizio
                    </InputLabel>
                    <Select
                        labelId="servicesIdLabel"
                        name="serviceId"
                        defaultValue={services[0].id}
                        inputProps={{
                            'data-testid': 'service',
                        }}
                    >
                        {services.map((service: Service) => (
                            <MenuItem key={service.id} value={service.id}>
                                {service.name}
                            </MenuItem>
                        ))}
                    </Select>
                </FormControl>

                <FormControl variant="outlined" className={classes.field} required>
                    <InputLabel className={classes.label} id="employeeIdLabel">
                        Assegnato a
                    </InputLabel>
                    <Select
                        labelId="employeeIdLabel"
                        name="employeeId"
                        value={employeeId}
                        onChange={handleChangeEmployeeId}
                        inputProps={{
                            'data-testid': 'employee',
                        }}
                    >
                        {employees.map((employee: Employee) => (
                            <MenuItem key={employee.id} value={employee.id}>
                                {employee.name}
                            </MenuItem>
                        ))}
                    </Select>
                </FormControl>

                <TextField
                    className={classes.field}
                    variant="outlined"
                    autoComplete="off"
                    name="notes"
                    label="Note"
                    multiline
                    rows={4}
                    inputProps={{
                        'data-testid': 'notes',
                    }}
                />

                <div className={classes.submitButton}>
                    <Button
                        type="submit"
                        color="primary"
                        variant="contained"
                        disabled={loading}
                        startIcon={loading ? <CircularProgress size={20} /> : null}
                        data-testid="submit-button"
                    >
                        {loading ? 'Caricamento' : 'Aggiungi'}
                    </Button>
                    <Button variant="contained" color="primary" disabled={loading} onClick={() => history.goBack()}>
                        Annulla
                    </Button>
                </div>
            </form>
        </div>
    );
};

export default React.memo(
    AddReservation,
    (prevProps, nextProps) => prevProps.location.search === nextProps.location.search,
);
