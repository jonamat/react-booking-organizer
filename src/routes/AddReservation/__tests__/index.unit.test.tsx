/**
 * Unit test suite
 *
 * Target: AddReservation component
 * Tests:
 * - Responses handling
 * - Redirection
 * - User interaction interruption on loading
 * - Middleware function call
 * - Visual feedbacks
 * - Snapshot matching
 * - Functionality
 *
 * For data validation, API call and errors handling see {@link '../../../utils/addReservation'}
 */

// Test resources
import React from 'react';
import renderer from 'react-test-renderer';
import { StylesProvider, createGenerateClassName } from '@material-ui/core';
import { render, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import DateFnsUtils from '@date-io/date-fns';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';

// Test configuration
import { addReservation } from '../../../utils';
import { __mockCustomers, __mockServices, __mockEmployees, __mockReservations } from '../../../__mocks__/react-redux';

// Mocked modules
import { toast } from 'react-toastify';

// Component to test
import AddReservation from '..';

// WARN: It could change based on timezone
const parsedDate = new Date('1 Jan 2040 10:00 GMT+0100');

/* -------------------------------------------------------------------------- */
/*                                    Mocks                                   */
/* -------------------------------------------------------------------------- */
jest.mock('../../../utils', () => ({
    ...(jest.requireActual('../../../utils') as Record<string, any>),
    addReservation: jest.fn(() => Promise.resolve({ id: 'fakeID', date: new Date('1 Jan 2040 10:00 GMT+0100') })),
    generateClosestValidDate: () => new Date('1 Jan 2020 10:00 GMT+0100'),
}));
jest.mock('../../../components/HolidaysField', () => () => null);
jest.mock('../../../components/ShiftsField', () => () => null);
const spyTSuccess = jest.spyOn(toast, 'success');
const spyTError = jest.spyOn(toast, 'error');
const history = {
    push: jest.fn(),
};

/* -------------------------------------------------------------------------- */
/*                              Bridge operations                             */
/* -------------------------------------------------------------------------- */

beforeEach(() => {
    jest.clearAllMocks();
});

/* -------------------------------------------------------------------------- */
/*                               Render wrapper                               */
/* -------------------------------------------------------------------------- */

/**
 * Render the component with the custom configuration
 * @param props Component props
 */
const renderComponent = (props?) => {
    return render(
        <MuiPickersUtilsProvider utils={DateFnsUtils}>
            <AddReservation history={history} location={{}} {...props} />
        </MuiPickersUtilsProvider>,
    );
};

/* -------------------------------------------------------------------------- */
/*                                    Tests                                   */
/* -------------------------------------------------------------------------- */

describe('AddReservation component', () => {
    describe('Upload system', () => {
        it('calls the db middleware function with correct user inputs', () => {
            const { getByTestId } = renderComponent();

            // Customer input
            fireEvent.input(getByTestId('customer'), { target: { value: __mockCustomers[0].name } });
            fireEvent.keyDown(getByTestId('customer'), { key: 'Enter', code: 'Enter' });

            // Date input and validation
            fireEvent.change(getByTestId('date'), { target: { value: '01/01/2040 10:00' } });
            fireEvent.blur(getByTestId('date'));

            // Other text fields
            fireEvent.input(getByTestId('service'), { target: { value: __mockServices[0].id } });
            fireEvent.input(getByTestId('employee'), { target: { value: __mockEmployees[0].id } });
            fireEvent.input(getByTestId('notes'), { target: { value: 'notes' } });

            fireEvent.click(getByTestId('submit-button'));

            expect(addReservation as jest.Mock).toBeCalledTimes(1);

            expect(Array.from((addReservation as jest.Mock).mock.calls[0][0].entries())).toEqual([
                ['serviceId', __mockServices[0].id],
                ['employeeId', __mockEmployees[0].id],
                ['notes', 'notes'],
            ]);
            expect((addReservation as jest.Mock).mock.calls[0][1]).toEqual(__mockCustomers[0]);
            expect((addReservation as jest.Mock).mock.calls[0][2]).toEqual(parsedDate);
            expect((addReservation as jest.Mock).mock.calls[0][3]).toEqual({
                employees: __mockEmployees,
                services: __mockServices,
                reservations: __mockReservations,
            });
            expect((addReservation as jest.Mock).mock.calls[0][4]).toBeUndefined();
        });

        it('prevents sending of invalid dates', async () => {
            const { getByTestId, queryAllByText } = renderComponent();

            // Date input and validation
            fireEvent.change(getByTestId('date'), { target: { value: '99/99/2090 70:00' } }); // invalid day month hour
            fireEvent.blur(getByTestId('date'));

            expect((await queryAllByText('Data non valida')).length).toBeTruthy();

            // Date input and validation
            fireEvent.change(getByTestId('date'), { target: { value: '10/10/2090 07:00' } });
            fireEvent.blur(getByTestId('date'));

            expect((await queryAllByText('Data non valida')).length).toBeFalsy();
        });

        it('parses and loads queries in url', () => {
            const { getByTestId } = renderComponent({
                location: {
                    search: `customerId=${__mockCustomers[0].id}&employeeId=${__mockEmployees[0].id}&date=2040-01-01&hour=10:00`,
                },
            });

            expect((getByTestId('customer') as HTMLInputElement).value).toBe(
                `${__mockCustomers[0].name} ${__mockCustomers[0].surname}`,
            );
            expect((getByTestId('employee') as HTMLInputElement).value).toBe(__mockEmployees[0].id);
            expect((getByTestId('date') as HTMLInputElement).value).toBe('01/01/2040 10:00');
        });
    });

    describe('User experience', () => {
        it('Autocomplete field sorts customers properly', () => {
            const { getByTestId } = renderComponent();
            fireEvent.click(getByTestId('customer'));
            fireEvent.input(getByTestId('customer'), { target: { value: __mockCustomers[0].name.slice(0, 5) } });
            fireEvent.keyDown(getByTestId('customer'), { key: 'Enter', code: 'Enter' });
            expect(getByTestId('customer')).toHaveValue(__mockCustomers[0].name + ' ' + __mockCustomers[0].surname);
        });

        it('redirects user on success', async () => {
            const { getByTestId } = renderComponent();

            // Customer input
            fireEvent.input(getByTestId('customer'), { target: { value: __mockCustomers[0].name } });
            fireEvent.keyDown(getByTestId('customer'), { key: 'Enter', code: 'Enter' });

            // Date input and validation
            fireEvent.change(getByTestId('date'), { target: { value: '01/01/2040 10:00' } });
            fireEvent.blur(getByTestId('date'));

            // Other text fields
            fireEvent.input(getByTestId('service'), { target: { value: __mockServices[0].id } });
            fireEvent.input(getByTestId('employee'), { target: { value: __mockEmployees[0].id } });

            fireEvent.click(getByTestId('submit-button'));

            await waitFor(() => {
                expect(history.push).toBeCalledTimes(1);
                expect(history.push).toBeCalledWith(`/reservations?date=2040-01-01`);
            });
        });

        it('calls toast on success', async () => {
            const { getByTestId } = renderComponent();

            // Customer input
            fireEvent.input(getByTestId('customer'), { target: { value: __mockCustomers[0].name } });
            fireEvent.keyDown(getByTestId('customer'), { key: 'Enter', code: 'Enter' });

            // Date input and validation
            fireEvent.change(getByTestId('date'), { target: { value: '01/01/2040 10:00' } });
            fireEvent.blur(getByTestId('date'));

            // Other text fields
            fireEvent.input(getByTestId('service'), { target: { value: __mockServices[0].id } });
            fireEvent.input(getByTestId('employee'), { target: { value: __mockEmployees[0].id } });

            fireEvent.click(getByTestId('submit-button'));

            await waitFor(() => {
                expect(spyTSuccess).toBeCalledTimes(1);
            });
        });

        it('calls toast on error', async () => {
            (addReservation as jest.Mock).mockImplementation(() => Promise.reject('some error'));

            const { getByTestId } = renderComponent();

            // Customer input
            fireEvent.input(getByTestId('customer'), { target: { value: __mockCustomers[0].name } });
            fireEvent.keyDown(getByTestId('customer'), { key: 'Enter', code: 'Enter' });

            // Date input and validation
            fireEvent.change(getByTestId('date'), { target: { value: '01/01/2040 10:00' } });
            fireEvent.blur(getByTestId('date'));

            // Other text fields
            fireEvent.input(getByTestId('service'), { target: { value: __mockServices[0].id } });
            fireEvent.input(getByTestId('employee'), { target: { value: __mockEmployees[0].id } });

            fireEvent.click(getByTestId('submit-button'));

            await waitFor(() => {
                expect(spyTError).toBeCalledTimes(1);
            });
        });

        it('disables submit button on loading', async () => {
            (addReservation as jest.Mock).mockImplementation(() => Promise.reject('some error'));

            const { getByTestId } = renderComponent();

            // Customer input
            fireEvent.input(getByTestId('customer'), { target: { value: __mockCustomers[0].name } });
            fireEvent.keyDown(getByTestId('customer'), { key: 'Enter', code: 'Enter' });

            // Date input and validation
            fireEvent.change(getByTestId('date'), { target: { value: '01/01/2040 10:00' } });
            fireEvent.blur(getByTestId('date'));

            // Other text fields
            fireEvent.input(getByTestId('service'), { target: { value: __mockServices[0].id } });
            fireEvent.input(getByTestId('employee'), { target: { value: __mockEmployees[0].id } });

            fireEvent.click(getByTestId('submit-button'));

            await waitFor(() => {
                expect(getByTestId('submit-button').hasAttribute('disabled')).toBeFalsy();
            });
        });
    });

    it('matches snapshot', () => {
        const snapshot = renderer
            .create(
                <MuiPickersUtilsProvider utils={DateFnsUtils}>
                    <StylesProvider generateClassName={createGenerateClassName({ disableGlobal: true })}>
                        {/* @ts-ignore */}
                        <AddReservation />
                    </StylesProvider>
                </MuiPickersUtilsProvider>,
            )
            .toJSON();
        expect(snapshot).toMatchSnapshot();
    });
});
