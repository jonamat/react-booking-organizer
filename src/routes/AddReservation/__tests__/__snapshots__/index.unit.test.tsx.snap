// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReservation component matches snapshot 1`] = `
<div
  className="jss1"
  data-testid="add-reservation"
>
  <h6
    className="jss8 jss18"
  >
    Aggiungi appuntamento
  </h6>
  <hr
    className="jss38 jss2"
  />
  <form
    className="jss3"
    data-testid="form"
    onSubmit={[Function]}
  >
    <div
      aria-expanded={false}
      aria-owns={null}
      className="jss45 jss5 jss51 jss50"
      onClick={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="combobox"
    >
      <div
        className="jss70 jss69 jss73"
      >
        <div
          className="jss89 jss74 jss52 jss99 jss90 jss94 jss79"
          onClick={[Function]}
        >
          <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-controls={null}
            aria-invalid={false}
            autoCapitalize="none"
            autoComplete="off"
            autoFocus={false}
            className="jss100 jss84 jss53 jss54 jss105 jss88"
            data-testid="customer"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onMouseDown={[Function]}
            placeholder="Cliente"
            required={false}
            spellCheck="false"
            type="text"
            value=""
          />
          <div
            className="jss55"
          >
            <button
              aria-label="Clear"
              className="jss116 jss107 jss56 jss4"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              title="Clear"
              type="button"
            >
              <span
                className="jss115"
              >
                <svg
                  aria-hidden={true}
                  className="jss119 jss126"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
            </button>
            <button
              aria-label="Open"
              className="jss116 jss107 jss58 jss4"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              title="Open"
              type="button"
            >
              <span
                className="jss115"
              >
                <svg
                  aria-hidden={true}
                  className="jss119"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <fieldset
            aria-hidden={true}
            className="jss128 jss83"
            style={
              Object {
                "paddingLeft": 8,
              }
            }
          >
            <legend
              className="jss129"
              style={
                Object {
                  "width": 0.01,
                }
              }
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="jss70 jss69 jss5"
    >
      <label
        className="jss144 jss132 jss138 jss141 jss140 jss143 jss149"
        data-shrink={true}
      >
        Data *
      </label>
      <div
        className="jss89 jss74 jss90 jss94 jss79"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="jss100 jss84 jss105 jss88"
          data-testid="date"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value="01/01/2020 10:00"
        />
        <div
          className="jss152 jss155"
        >
          <button
            className="jss116 jss107"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="jss115"
            >
              <svg
                aria-hidden={true}
                className="jss119"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
          </button>
        </div>
        <fieldset
          aria-hidden={true}
          className="jss128 jss83"
        >
          <legend
            className="jss130 jss131"
          >
            <span>
              Data *
            </span>
          </legend>
        </fieldset>
      </div>
      
    </div>
    <div
      className="jss70 jss5"
    >
      <label
        className="jss144 jss132 jss6 jss138 jss141 jss140 jss143 jss149 jss150 jss136"
        data-shrink={true}
        id="servicesIdLabel"
      >
        Servizio
        <span
          aria-hidden={true}
          className="jss151 jss137"
        >
           
          *
        </span>
      </label>
      <div
        className="jss89 jss74 jss90"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="servicesIdLabel mui-component-select-serviceId"
          className="jss178 jss179 jss182 jss181 jss100 jss84"
          id="mui-component-select-serviceId"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          fake service 1
        </div>
        <input
          aria-hidden={true}
          className="jss188"
          data-testid="service"
          name="serviceId"
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={true}
          tabIndex={-1}
          value="service-FAKE-ID1"
        />
        <svg
          aria-hidden={true}
          className="jss119 jss184 jss187"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="jss128 jss83"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="jss129"
            style={
              Object {
                "width": 0.01,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="jss70 jss5"
    >
      <label
        className="jss144 jss132 jss6 jss138 jss141 jss140 jss143 jss149 jss150 jss136"
        data-shrink={true}
        id="employeeIdLabel"
      >
        Assegnato a
        <span
          aria-hidden={true}
          className="jss151 jss137"
        >
           
          *
        </span>
      </label>
      <div
        className="jss89 jss74 jss90"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="employeeIdLabel mui-component-select-employeeId"
          className="jss178 jss179 jss182 jss181 jss100 jss84"
          id="mui-component-select-employeeId"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Employee name 1
        </div>
        <input
          aria-hidden={true}
          className="jss188"
          data-testid="employee"
          name="employeeId"
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={true}
          tabIndex={-1}
          value="employee-FAKE-ID1"
        />
        <svg
          aria-hidden={true}
          className="jss119 jss184 jss187"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="jss128 jss83"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="jss129"
            style={
              Object {
                "width": 0.01,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="jss70 jss69 jss5"
    >
      <label
        className="jss144 jss132 jss138 jss141 jss143"
        data-shrink={false}
      >
        Note
      </label>
      <div
        className="jss89 jss74 jss90 jss97 jss82"
        onClick={[Function]}
      >
        <textarea
          aria-invalid={false}
          autoComplete="off"
          autoFocus={false}
          className="jss100 jss84 jss102 jss86"
          data-testid="notes"
          disabled={false}
          name="notes"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          rows={4}
        />
        <fieldset
          aria-hidden={true}
          className="jss128 jss83"
        >
          <legend
            className="jss130"
          >
            <span>
              Note
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="jss7"
    >
      <button
        className="jss116 jss193 jss201 jss202"
        data-testid="submit-button"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="jss194"
        >
          Aggiungi
        </span>
      </button>
      <button
        className="jss116 jss193 jss201 jss202"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="jss194"
        >
          Annulla
        </span>
      </button>
    </div>
  </form>
</div>
`;
